Improve expiration of tiered objects

krisis opened the pull request:
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
### User facing changes
- Added a new configuration subsystem (`ilm`) to manage number of workers dedicated to expiration and tiering jobs.
```
# To get the current configuration
mc admin config get ALIAS ilm

# To set number of workers for expiration jobs (default is 100)
mc admin config set ALIAS ilm expiration_workers=50
```
- Prometheus Metrics
```
- # HELP minio_node_ilm_expiry_num_workers Number of workers expiring object versions currently
  # TYPE minio_node_ilm_expiry_num_workers gauge

- # HELP minio_node_ilm_expiry_missed_tasks Number of object version expiry missed due to busy system
  # TYPE minio_node_ilm_expiry_missed_tasks counter

(Internal)
- # HELP minio_node_ilm_expiry_missed_freeversions Number of free versions expiry missed due to busy system
 # TYPE minio_node_ilm_expiry_missed_freeversions counter

- # HELP minio_node_ilm_expiry_missed_tierjournal_tasks Number of tier journal entries cleanup missed due to busy system
  # TYPE minio_node_ilm_expiry_missed_tierjournal_tasks counter


```
- Use a shared worker pool for all ILM expiry tasks
- Free version cleanup no longer runs in the scanner goroutine
- Add a free version only if removing the remote object fails
- Add ilm expiry metrics to node namespace

## How to test this PR?


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

Commit:
Improve expiration of tiered objects

- Use a shared worker pool for all ILM expiry tasks
- Free version cleanup executes in a separate goroutine
- Add a free version only if removing the remote object fails
- Add ilm expiry metrics to node namespace
- Move tier journal tasks to expiryState
- Remove unused on-disk journal for tiered objects pending deletion

Commit:
Move to a workerpool model similar to replication

- Distribute expiry tasks across workers such that expiry of versions of
  the same object serialized
- Ability to resize worker pool without server restart

shtripat commented on cmd/xl-storage-free-version_test.go:
```diff
@@ -228,3 +229,55 @@ func TestFreeVersion(t *testing.T) {
+			}},
+		},
+		MarkDeleted: false,
+		// DeleteMarkerReplicationStatus: "",
```
Remove commented if not needed

vadmeste commented on internal/config/api/help.go:
```diff
@@ -80,6 +80,12 @@ var (
 		},
+		config.HelpKV{
+			Key:         apiExpiryWorkers,
+			Description: `set the number of expiry workers` + defaultHelpPostfix(apiExpiryWorkers),
```
is this per node ?

krisis replied:
Yes, it is.

vadmeste commented on cmd/config-current.go:
```diff
@@ -711,6 +713,18 @@ func applyDynamicConfigForSubSys(ctx context.Context, objAPI ObjectLayer, s conf
+		if err != nil {
+			return fmt.Errorf("Unable to apply ilm config: %w", err)
+		}
+		if globalTransitionState != nil {
```
can you add a log when this is nil ?

vadmeste commented on cmd/config-current.go:
```diff
@@ -711,6 +713,18 @@ func applyDynamicConfigForSubSys(ctx context.Context, objAPI ObjectLayer, s conf
+		if globalTransitionState != nil {
+			globalTransitionState.UpdateWorkers(ilmCfg.TransitionWorkers)
+		}
+		if globalExpiryState != nil {
```
Can you add a log when this is nil ?

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -105,96 +102,266 @@ type expiryTask struct {
+// object corresponding to a 'replaced' object versions. This applies only to
+// non-versioned or version suspended buckets.
+func (es *expiryState) enqueueTierJournalEntry(je jentry) {
+	wrkr := es.getWorkerCh(je.OpHash())
```
can you elaborte more why we use a hash here ?

klauspost replied:
Probably the same as replication where we don't want different versions of the same object trying to update at the same time.

krisis replied:
Yes, it is for the same reason as we do this for replication workers.

vadmeste reviewed (commented):
Few comments

klauspost commented on cmd/bucket-lifecycle.go:
```diff
@@ -105,96 +102,266 @@ type expiryTask struct {
 // NewerNoncurrentVersions limit for expiry.
 func (es *expiryState) enqueueByNewerNoncurrent(bucket string, versions []ObjectToDelete, lcEvent lifecycle.Event) {
+	task := newerNoncurrentTask{bucket: bucket, versions: versions, event: lcEvent}
+	wrkr := es.getWorkerCh(task.OpHash())
```
This and probably similar workers are racy.

The channel returned by `es.getWorkerCh` has no protection against "downscaling" - meaning the channel is closed by `ResizeWorkers` before you can write to it. This is a crash bug.

The best idea I have is to make the "close"  a nil entry instead of actually closing.

klauspost commented on cmd/bucket-lifecycle.go:
```diff
@@ -105,96 +102,266 @@ type expiryTask struct {
+			if !ok {
+				return
+			}
+			switch v := v.(type) {
```
Log unexpected types.

klauspost commented on cmd/bucket-lifecycle.go:
```diff
@@ -105,96 +102,266 @@ type expiryTask struct {
 
-func initBackgroundExpiry(ctx context.Context, objectAPI ObjectLayer) {
-	globalExpiryState = newExpiryState()
+func (es *expiryState) getWorkerCh(h uint64) chan<- expiryOp {
```
Here is a lockless (for everything but resize) implementation:

```
// expiryState manages all ILM related expiration activities.
type expiryState struct {
	mu      sync.RWMutex
	workers atomic.Pointer[[]chan expiryOp]
}

func (es *expiryState) ResizeWorkers(n int) {
	// Lock to avoid multiple resizes to happen at the same time.
	es.mu.Lock()
	defer es.mu.Unlock()
	var workers []chan expiryOp
	if v := es.workers.Load(); v != nil {
		// Copy to new array.
		workers = append(workers, *v...)
	}

	if n == len(workers) || n < 1 {
		return
	}

	for len(workers) < n {
		input := make(chan expiryOp, 10000)
		workers = append(workers, input)
		go es.Worker(input)
		es.stats.workers.Add(1)
	}

	for len(workers) > n {
		worker := workers[len(workers)-1]
		workers = workers[:len(workers)-1]
		worker <- expiryOp(nil)
		es.stats.workers.Add(-1)
	}
	// Atomically replace workers.
	es.workers.Store(&workers)
}

func (es *expiryState) getWorkerCh(h uint64) chan<- expiryOp {
	w := es.workers.Load()
	if w == nil || len(*w) == 0 {
		return nil
	}
	workers := *w
	return workers[h%uint64(len(workers))]
}

// PendingTasks returns the number of pending ILM expiry tasks.
func (es *expiryState) PendingTasks() int {
	w := es.workers.Load()
	if w == nil || len(*w) == 0 {
		return 0
	}
	var tasks int
	for _, wrkr := range *w {
		tasks += len(wrkr)
	}
	return tasks
}

````

krisis replied:
Thank you for the detailed review and the alternate approach posted above! Now, I see the race between `ResizeWorkers` and any of the `enqueue*` functions. I shall refresh my PR with these changes.

Commit:
Address review comments

- Make scaling down of expiryState workers concurrency safe; Thanks
  @klauspost

- Add error logs when expiryState and transitionState are not
  initialized (yet)

Commit:
metrics: Add missed tier journal entry tasks

vadmeste reviewed (approved):
LGTM

Commit:
Initialize ILM worker pool after object layer
